src/OpenTelemetry.AutoInstrumentation.Native/cor_profiler.cpp

```diff
@@ -2494,14 +2494,47 @@ HRESULT CorProfiler::CallTarget_RewriterCallback(RejitHandlerModule* moduleHandl
             &wrapper_type_def);
         if (FAILED(hr) || wrapper_type_def == mdTypeDefNil)
         {
-            Logger::Warn("*** CallTarget_RewriterCallback() Failed for: ", caller->type.name, ".", caller->name,
+            Logger::Error("*** CallTarget_RewriterCallback() Failed for: ", caller->type.name, ".", caller->name,
                         "() integration type not found on the managed profiler module HRESULT=", HResultStr(hr),
                         " IntegrationType=", wrapper.type_name);
             return S_FALSE;
         }
 
-        // TODO: Use the method signature to validate also the method, tracking via
-        // https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/issue/1499 
+        // CallTarget instrumentation doesn't inject calls to the instrumentation methods via IL rewrite.
+        // It injects the OpenTelemetry.AutoInstrumentation.CallTarget.CallTargetInvoker, written in managed code,
+        // that uses reflection to find the expected instrumentation methods on the instrumentation wrapper type.
+        // If the wrapper type doesn't have any of the expected instrumentation methods "nothing happens", but,
+        // the JIT code of the targeted method is modified anyway. To avoid injecting instrumentation that does
+        // nothing and give a clear error message the code below ensures that at least one of the expected methods is
+        // implemented on the wrapper type.
+        static const LPCWSTR expected_wrapper_methods[] = { WStr("OnMethodBegin"), WStr("OnMethodEnd"), WStr("OnAsyncMethodEnd") };
+        bool found_wrapper_method = false;
+        for (auto i = 0; i < sizeof(expected_wrapper_methods) / sizeof(LPCWSTR); i++)
```

Comment:
Any chance to use for each loops to avoid calculation of array length?
https://www.w3schools.com/cpp/trycpp.asp?filename=demo_array_loop_foreach

Reply:
[73f67ae](https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/pull/1520/commits/73f67ae9744b3a2d54a39630c809a122f7d513e2)